Web app

danil-topc opened the pull request:
Init backend with api for features implemented in cli

matejak commented:
Thank you for the PR, I have a problem getting this to run, and as I don't have experience with vue and just a little bit of XP with node.js, I stopped here:

When in `src/web/teap/frontend`, I run `npm install` and it seems to work more or less fine.
However, when I run `npm run dev` (or any other command), I get an error:

```
$ npm run dev

> frontend@1.0.0 dev .../git/teap/src/web/teap/frontend
> webpack-dev-server --inline --progress --config build/webpack.dev.conf.js

internal/modules/cjs/loader.js:615
    throw err;
    ^

Error: Cannot find module '...j/git/teap/src/web/teap/frontend/build/webpack.dev.conf.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:613:15)
    at Function.Module._load (internal/modules/cjs/loader.js:539:25)
    at Module.require (internal/modules/cjs/loader.js:667:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at requireConfig (.../git/teap/src/web/teap/frontend/node_modules/webpack/bin/convert-argv.js:97:18)
    at .../git/teap/src/web/teap/frontend/node_modules/webpack/bin/convert-argv.js:104:17
    at Array.forEach (<anonymous>)
    at module.exports _(.../git/teap/src/web/teap/frontend/node_modules/webpack/bin/convert-argv.js:102:15)_
    at Object.<anonymous> (.../git/teap/src/web/teap/frontend/node_modules/webpack-dev-server/bin/webpack-dev-server.js:223:50)
    at Module._compile (internal/modules/cjs/loader.js:738:30)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! frontend@1.0.0 dev: `webpack-dev-server --inline --progress --config build/webpack.dev.conf.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the frontend@1.0.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .../.npm/_logs/2019-02-18T02_43_11_522Z-debug.log
```

danil-topc commented:
@matejak my fault, sorry. Didn't notice `build/` in global .gitignore, which is needed for webpack server. Fixed.
Also added simple Dockerfile to build and serve static with node http-server

matejak commented:
I can confirm that it works. I will take a more detailed look at it tomorrow to understand what is it doing. Thanks!

matejak commented:
So I took a look at your contribution, and it seems somehow over-engineered for the purpose. As I am not familiar with this approach, I understand that it may be worth it in the long run; could you please point us to a resource on

1. How this content was generated, and
2. what the web development workflow looks like?

You have mentioned using cookiecutter, but that project has many possible project templates.

danil-topc commented:
@matejak as we discussed previously, in the last commit 71454d70a32ce83816e92082557126c9b9c2578c I put existing web code base to [cookiecutter-flask template](https://github.com/sloria/cookiecutter-flask) and deleted redundant features from there (as webpack, flask static pages, views and forms). 
It looks a bit complex from first sight, but it implements good way to maintain flask app, as app factory used by wsgi, and easily used for testing with different config. Also it implements configuration using env variables as in [12-factor-app](https://12factor.net/config), adds relational database support  (I think we will need it at some point anyway) basic User, Role models, migration system with Flask-Migrate. As well it implements pytest with basic tests, caching, debug toolbar.

As for development workflow, it uses modular applications with [blueprints](http://flask.pocoo.org/docs/1.0/blueprints/), so we will be able to simplify development and make each part of the project, such as nextcloud, edap in different application. For each db change we will generate migrations, so everybody on the project will be easy to follow database state. And of course tests.

matejak commented on src/web/teap/teap/compat.py:
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
```
We can get rid of this, as Python2 compatibility is not a goal.

danil-topc replied:
agree, it was by default in template, so will clean it up

matejak commented on src/web/teap/teap/nextcloud/api.py:
```diff
@@ -0,0 +1,52 @@
+    def nextcloud(self):
+        """ Get nextcloud instance """
+        # tmp mock nextcloud credentials
+        url = 'http://localhost:8080'
```
Obviously, these values are candidates to make configurable e.g. by using env variables.

danil-topc replied:
I think it will be attached to user somehow. As I understand platform can be used for different nextcloud instances, or it will be deployed for the only one?

matejak replied:
Great question! The current use case is one instance. Imagine an enterprisey setup featuring e.g. Nextcloud, Rocket.Chat, Gitlab, etc. - you will have one instance for each (at least at the beginning), but at the same time there will be those multiple services that have to be managed.
The idea is that the service will have admin access, as only admins will be allowed to get inside.

danil-topc replied:
Thanks for clarification! I moved values to config. 
We will probably need to attach username, password for every user separately in the future, but for now left it in the config as well.

Off topic: previously I thought that it should be some kind of service where anybody can register, add their own instances of Nextcloud, Gitlab, whatever, and manage them. But it probably looks more organic as a part of enterprise build, not as SaaS.

matejak commented on src/web/teap/teap/nextcloud/api.py:
```diff
@@ -0,0 +1,52 @@
+        return nxc
+
+
+class UserViewSet(NextCloudMixin,
```
Just a clarification - TEAP won't need a REST API, it will use REST APIs of other services, while being a user-friendly "command center". Disregard this comment if you intend to feed output of these methods to client-side JS to reformat them.

danil-topc replied:
Yes, it is used on our vue.js front-end already

matejak commented on src/web/teap/requirements/dev.txt:
```diff
@@ -0,0 +1,17 @@
+# Everything the developer needs in addition to the production requirements
```
I don't like hardcoding those version numbers here in the repository, could you do a trick similar to nextcloud's `requirements.in`? Moreover, is `flake8` really a hard dependency?

danil-topc replied:
Ok, will refactor it to `requirements.in`. About `flake8` this is was by default in template, but only in development requirements, so won't be used in production, but still we can get rid of this anyways.

matejak replied:
Sure, it looks like that it comes from some kind of template. However, devs can use whatever they want for coding support, so let's keep this minimal. I dislike concerte versions outside of releases, as they require constant maintenance.

matejak commented on src/web/teap/requirements/prod.txt:
```diff
@@ -0,0 +1,38 @@
+# Everything needed in production
```
Ditto here, some dependencies are second or third order. Let's have a `prod.in` with the essential requirements if it is possible.

danil-topc replied:
Yes, will do

matejak commented:
Thank you for clarification! I have made some comments, and I have following points:

- We will make use of databases at some point, but authenticating users will be accomplished by querying a LDAP directory, probably using the EDAP project. You don't have to take the SQLAlchemy DB  "scaffolding" out, but we will use it in a slightly different way.
- The directory structure is a bit confusing, we have `src/web` as web stuff root (totally OK), and then we have `teap` directory, which has another `teap` subdirectory. I have to say that I would take the first `teap` out completely, and I would consider renaming the second `teap` to `backend`, as it is exactly the Python backend. What do you think of this? However, I am in favor of adhering to conventions and to best practices, so if it is the expected way of doing things, let it stay like that.
- I have noticed that `backend/api.py`, and `teap/nextcloud/api.py` files are basically duplicates. Could you look into this? If you decide to remove any of them, consider using the `git rebase -i` command, that will keep the commit history of this PR clean - so far, it is great for this PR, you did a swell job!

danil-topc commented:
@matej thanks for review! 

-  Ok, then I will just leave it as is, until we will need it.
-  Agree, was just thinking to rebase it that way, will do.
-  Yes, forgot to add deleted files to commit with switching to different project structure, just deleted it in last commit few minutes ago, will try to clean it out of history.

danil-topc commented:
Also how do you think, after making all suggested changes we should merge it and continue work in another branch or I should also add some front-end styling and improve UI in this branch?

matejak commented:
I have checked your commits again, and I ran into an import issue:  I got collision of the `nextcloud` subpackage and `nextcloud` as the installed Nextcloud API package. I am somehow surprised by that as Python3 is supposed to use the system package first, but it may also be a venv-related issue. Consider renaming the package, as it would increase the robustness.

danil-topc commented:
> I got collision of the `nextcloud` subpackage and `nextcloud` as the installed Nextcloud API package
This is because when we run application, the directory containing the script will be placed in the beginning of `sys.path`. 
According to [docs](https://docs.python.org/3/tutorial/modules.html#the-module-search-path):
> After initialization, Python programs can modify sys.path. The directory containing the script being run is placed at the beginning of the search path, ahead of the standard library path. This means that scripts in that directory will be loaded instead of modules of the same name in the library director.
But that's quite unobvious behavior, agree.  

But that nextcloud subpackage is not in the root, it placed inside backend folder, so we don't have to face any collisions during normal run of application from root directory (`web/teap`).

danil-topc commented:
@matejak also didn't add nextcloud to requirements, because not able to install it via ` pip install git+git://github.com/EnterpriseyIntranet/nextcloud-API#egg=nextcloud`, it falls with exception:
``` setup() arguments must *always* be /-separated paths relative to the
    setup.py directory, *never* absolute paths```
Could you, please, suggest best practices fixing that ?

matejak commented:
Great, I have checked it out and it seems to work quite nicely. Could you please update all READMEs, or discard them if you feel that they are not needed?
For reference, I have built the frontend, and as a second step, I have created my own virtualenv, so I have invoked the backend using
```
DATABASE_URL=sqlite:///temp.sqlite SECRET_KEY=... ... python -m flask run
```
in `teap/src/web`. Is this the expected way to launch the web app?

danil-topc commented:
@matejak strange, because I'm still unable to install it with that command.
Pushed minor updates for READMEs.  
Yes, that's expected way to run it. Also, for development you can use `src/web/.env` file for settings variables and `npm run dev` (by default runs on 8081 port) for hot-reload debug server on front-end, flask will redirect all requests to it, when `FLASK_DEBUG=True`.

matejak commented:
Great, this is good for now. More specs will be on the way, and I will try to deploy this somewhere as well.
